TopPlaces/TopPlaces/Controllers/DetailsPhotoViewController.m

```diff
@@ -0,0 +1,208 @@
+// Created by Gennadi Iosad.
+
+#import "DetailsPhotoViewController.h"
+
+#import "PhotosHistory.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface DetailsPhotoViewController() <UIScrollViewDelegate>
+
+@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;
+@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;
+@property (weak, nonatomic) IBOutlet UILabel *noImageLoadedView;
+@property (strong, nonatomic) UIImageView *imageView;
+@property (strong, nonatomic) NSURLSessionTask *downloadTask;
+
+@end
+
+@implementation DetailsPhotoViewController
+
+#pragma mark -
+#pragma mark UIViewController overrides
+#pragma mark -
+
+- (void)viewDidLoad {
+  [super viewDidLoad];
+  self.scrollView.delegate = self;
+  UITapGestureRecognizer *doubleTapRecognizer =
+      [[UITapGestureRecognizer alloc] initWithTarget:self
+                                              action:@selector(scrollViewDoubleFingerTapped:)];
+  doubleTapRecognizer.numberOfTapsRequired = 2;
+  doubleTapRecognizer.numberOfTouchesRequired = 1;
+  [self.scrollView addGestureRecognizer:doubleTapRecognizer];
+  [self loadImage];
+  NSLog(@"DetailsPhotoViewController::viewDidLoad self %p", self);
+}
+
+
+- (void)scrollViewDoubleFingerTapped:(UITapGestureRecognizer*)recognizer {
+  //if already zoomed out - zoom in a bit
+  if (self.scrollView.zoomScale == self.scrollView.minimumZoomScale) {
+    [self.scrollView setZoomScale:self.scrollView.zoomScale*2 animated:YES];
+  } else {
+    //zoom out
+    [self.scrollView setZoomScale:self.scrollView.minimumZoomScale animated:YES];
+  }
+}
+
+
+- (void)viewDidAppear:(BOOL)animated {
+  NSLog(@"DetailsPhotoViewController::viewDidAppear");
+  [super viewDidAppear:animated];
+}
+
+
+- (void)viewDidLayoutSubviews {
+  NSLog(@"DetailsPhotoViewController::viewDidLayoutSubviews");
+  [super viewDidLayoutSubviews];
+
+  if (self.imageView) {
+    [self aspectFitImage];
+  }
+}
+
+- (void) viewDidDisappear:(BOOL)animated {
+  [super viewDidDisappear:animated];
+  NSLog(@"DetailsPhotoViewController::viewDidDisappear");
+  [self.downloadTask cancel];
+}
+
+#pragma mark -
+#pragma mark UIScrollViewDelegate impl
+#pragma mark -
+
+- (nullable UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
+  return self.imageView;
+}
+
+- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
+  // The scroll view has zoomed, so you need to re-center the contents
+  [self centerScrollViewContents];
+}
+
+#pragma mark -
+#pragma mark aux logic
+#pragma mark -
+
+- (void)loadImage {
+  //in ipad info can be missing
+  if (!self.photoInfo) {
+    return;
+  }
+
+  [PhotosHistory addPhotoInfo:self.photoInfo];
+  
+  self.noImageLoadedView.hidden = YES;
+  [self.spinner startAnimating];
+
+  __weak DetailsPhotoViewController *weakSelf = self;
+  self.downloadTask =
+      [self.placesPhotosProvider downloadPhoto:self.photoInfo
+                             completionHandler:^(UIImage *img, NSError *error) {
```

Comment:
Writing this completion handler inplace bloats the loadImage method, extract the completion handler to a separate method. Do the same for the fetching methods in the other VCs.